app/infrastructure/persistence/repositories/event_repo.py

```
@@ -55,10 +54,32 @@ async def get_last_event(self, subject_id: str, tenant_id: str) -> EventResult |
+            return {}
+        result = await self.db.execute(
+            select(Event)
+            .where(
+                Event.tenant_id == tenant_id,
+                Event.subject_id.in_(subject_ids),
+            )
+            .order_by(Event.subject_id, desc(Event.event_time), desc(Event.id))
+        )
+        rows = result.scalars().all()
+        # First row per subject_id is the latest (ordered by event_time desc, id desc).
+        out: dict[str, EventResult | None] = {sid: None for sid in subject_ids}
```

Ruff (0.15.0)

[convention] Unnecessary dict comprehension for iterable; use `dict.fromkeys` instead

Replace with `dict.fromkeys(iterable, value)`)

(C420)